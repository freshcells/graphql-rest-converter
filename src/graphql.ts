import _ from 'lodash'
import {
  DocumentNode,
  GraphQLSchema,
  visit,
  print,
  Kind,
  ASTNode,
  buildASTSchema,
  OperationDefinitionNode,
  GraphQLDirective,
} from 'graphql'
import { gql } from 'graphql-tag'
import { OpenAPIV3 } from 'openapi-types'
import { getVariablesFromPathTemplate } from './pathTemplate'
import {
  createFragmentMap,
  getDependencyClosure,
  getDirective,
  getDirectiveArguments,
  getFragmentDependencies,
  getReferencedFragments,
  isOperationDefinitionNode,
} from './graphqlUtils'
import { CustomProperties, OAType, OpenAPIGraphQLOperation } from './types'
import { getReferenceableFragments, GraphQLTypeToOpenAPITypeSchemaConverter } from './typeConverter'

const DIRECTIVE_DEFINITION = gql`
  input OAExternalDocsInput {
    url: String!
    description: String
  }

  enum HTTPVerb {
    POST
    PUT
    DELETE
  }

  enum ParameterSource {
    PATH
    QUERY
    HEADER
  }

  directive @OAQuery(
    path: String!
    tags: [String!]
    summary: String
    description: String
    externalDocs: OAExternalDocsInput
    deprecated: Boolean
  ) on QUERY

  directive @OAParam(
    in: ParameterSource
    deprecated: Boolean
    description: String
    name: String
  ) on VARIABLE_DEFINITION

  directive @OAMutation(
    path: String!
    tags: [String!]
    summary: String
    description: String
    externalDocs: OAExternalDocsInput
    deprecated: Boolean
    "defaults to POST"
    httpVerb: HTTPVerb
  ) on MUTATION

  directive @OABody(
    "Assume variable name if undefined"
    path: String
    deprecated: Boolean
    description: String
  ) on VARIABLE_DEFINITION

  directive @OAMap(to: String, from: String) on INPUT_FIELD_DEFINITION | VARIABLE_DEFINITION
`

enum OpenAPIDirectives {
  Query = 'OAQuery',
  Param = 'OAParam',
  Mutation = 'OAMutation',
  Body = 'OABody',
  Map = 'OAMap',
}

const graphqlQueryDirectiveDataToOpenAPIOperation = (requestConfig: OpenAPIQueryDirectiveData) => {
  const { tags, summary, description, externalDocs, deprecated } = requestConfig
  return {
    ...(tags?.length ? { tags } : {}),
    ...(typeof summary === 'string' ? { summary } : {}),
    ...(typeof description === 'string' ? { description } : {}),
    ...(typeof externalDocs?.url === 'string' ? { externalDocs } : {}),
    ...(deprecated ? { deprecated } : {}),
  }
}

const getOpenAPIParameters = (
  parameters: Omit<OpenAPIV3.ParameterObject, 'in'>[],
  path: string,
  variablesDirectiveData: Record<string, OpenAPIParamDirectiveData>
) => {
  const adjustedParameters: OpenAPIV3.ParameterObject[] = []
  const variableMap: Record<string, string> = {}

  const pathVariables = new Set(getVariablesFromPathTemplate(path))
  for (const parameter of parameters) {
    const variableName = parameter.name
    const variableDirectiveData = variablesDirectiveData[variableName]
    let parameterName = variableName
    const nameOverride = variableDirectiveData.name
    if (nameOverride) {
      parameterName = nameOverride
    }
    const variableDirectiveDataIn = variableDirectiveData.in?.toLowerCase()

    if (variableDirectiveDataIn) {
      if (pathVariables.has(parameterName) && variableDirectiveDataIn !== 'path') {
        throw new Error(
          `Location ${variableDirectiveDataIn} invalid for parameter ${parameterName} because it is part of the path`
        )
      }
      if (!pathVariables.has(parameterName) && variableDirectiveDataIn === 'path') {
        throw new Error(
          `Location ${variableDirectiveDataIn} invalid for parameter ${parameterName} because it is not part of the path`
        )
      }
      if (variableDirectiveDataIn === 'cookie') {
        throw new Error(`Unsupported parameter location cookie for parameter ${parameterName}`)
      }
      if (!['path', 'query', 'header'].includes(variableDirectiveDataIn)) {
        throw new Error(
          `Unknown parameter location ${variableDirectiveDataIn} for parameter ${parameterName}`
        )
      }
    }
    if (variableDirectiveDataIn && variableDirectiveDataIn === 'header') {
      parameterName = parameterName.toLowerCase()
    }
    if (parameterName !== variableName) {
      variableMap[parameterName] = variableName
    }

    const in_ = pathVariables.has(parameterName)
      ? 'path'
      : variableDirectiveDataIn && ['header', 'cookie'].includes(variableDirectiveDataIn)
      ? variableDirectiveDataIn
      : 'query'
    const deprecated = variableDirectiveData.deprecated
    const description = variableDirectiveData.description
    adjustedParameters.push({
      ...parameter,
      in: in_,
      name: parameterName,
      ...(nameOverride ? { [CustomProperties.VariableName]: parameter.name } : {}),
      ...(deprecated === true ? { deprecated } : {}),
      ...(description ? { description } : {}),
    })
  }

  return {
    parameters: adjustedParameters,
    variableMap,
  }
}

type OpenAPIQueryDirectiveData = {
  path: string
  tags?: [string]
  summary?: string
  description?: string
  externalDocs?: {
    url: string
    description?: string
  }
  deprecated?: boolean
}

type OpenAPIParamDirectiveData = {
  in?: string
  deprecated?: boolean
  description?: string
  name?: string
}

const createOpenAPIOperation = (
  operationId: string | null,
  operationSource: string,
  parameters: OpenAPIV3.ParameterObject[],
  responseSchema: OAType,
  queryDirectiveData: OpenAPIQueryDirectiveData
) => {
  const responses = {
    '200': {
      description: 'Success',
      content: {
        'application/json': {
          schema: responseSchema,
        },
      },
    },
    '400': {
      description: 'Invalid request',
      content: {
        'application/json': {},
      },
    },
  }
  return {
    ...(operationId !== null ? { operationId } : {}),
    [CustomProperties.Operation]: operationSource,
    ...graphqlQueryDirectiveDataToOpenAPIOperation(queryDirectiveData),
    parameters,
    responses,
  }
}

const getVariablesDirectiveData = (
  directive: GraphQLDirective,
  operation: OperationDefinitionNode
) => {
  const result: Record<string, OpenAPIParamDirectiveData> = {}
  for (const variableDefinition of operation.variableDefinitions || []) {
    const paramDirective = getDirective(variableDefinition, OpenAPIDirectives.Param)
    result[variableDefinition.variable.name.value] = paramDirective
      ? (getDirectiveArguments(directive, paramDirective) as OpenAPIParamDirectiveData)
      : {}
  }
  return result
}

export const getOpenAPIGraphQLOperations = (
  schema: GraphQLSchema,
  document: DocumentNode,
  customScalars?: (scalarTypeName: string) => OpenAPIV3.SchemaObject
) => {
  const result: Array<OpenAPIGraphQLOperation> = []

  const fragmentMap = createFragmentMap(document.definitions)

  const fragmentDependencies = getFragmentDependencies(fragmentMap)

  const referencableFragments = getReferenceableFragments(schema, fragmentMap, document)

  const typeConverter = new GraphQLTypeToOpenAPITypeSchemaConverter(
    schema,
    customScalars,
    fragmentMap,
    referencableFragments
  )

  const directiveSchema = buildASTSchema(DIRECTIVE_DEFINITION)

  for (const operation of document.definitions) {
    if (!isOperationDefinitionNode(operation)) {
      continue
    }

    const queryDirective = getDirective(operation, OpenAPIDirectives.Query)
    if (!queryDirective) {
      continue
    }

    const queryDirectiveData = getDirectiveArguments(
      directiveSchema.getDirective(OpenAPIDirectives.Query)!,
      queryDirective
    ) as OpenAPIQueryDirectiveData

    const variablesDirectiveData = getVariablesDirectiveData(
      directiveSchema.getDirective(OpenAPIDirectives.Param)!,
      operation
    )

    const operation_ = removeOpenAPIDirectives(operation)

    const operationFragmentDependencies = getDependencyClosure(
      getReferencedFragments(operation_),
      fragmentDependencies
    )
    const referencedFragments = Object.values(
      _.pickBy(fragmentMap, (v, k) => operationFragmentDependencies.has(k))
    )

    const singleOperationDocument = {
      kind: Kind.DOCUMENT,
      definitions: [operation_, ...referencedFragments],
    } as DocumentNode

    const operationId = operation_.name ? operation_.name.value : null

    const parametersSchema = typeConverter.parametersFromOperation(operation_)

    const responseSchema = typeConverter.responseSchemaFromOperation(operation_)

    const { parameters, variableMap } = getOpenAPIParameters(
      parametersSchema,
      queryDirectiveData.path,
      variablesDirectiveData
    )

    const operationSource = print(singleOperationDocument)

    const openAPIOperation = createOpenAPIOperation(
      operationId,
      operationSource,
      parameters,
      responseSchema,
      queryDirectiveData
    )

    const graphqlOpenAPIOperation = {
      openAPIOperation,
      httpMethod: OpenAPIV3.HttpMethods.GET,
      graphqlDocument: singleOperationDocument,
      variableMap,
      path: queryDirectiveData.path,
    }
    result.push(graphqlOpenAPIOperation)
  }

  return {
    operations: result,
    schemaComponents: typeConverter.getSchemaComponents(),
  }
}

const removeOpenAPIDirectives = <T extends ASTNode>(node: T): T => {
  return visit(node, {
    Directive: {
      enter(node) {
        if (Object.values(OpenAPIDirectives).includes(node.name.value as OpenAPIDirectives)) {
          return null
        }
      },
    },
  })
}
