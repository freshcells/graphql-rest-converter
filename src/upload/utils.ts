import { OperationDefinitionNode } from 'graphql/index.js'
import { DocumentNode } from 'graphql'

export const getAllVariablesFromDocuments = (docs: DocumentNode[]) => {
  return Object.fromEntries(
    docs.flatMap((document, index) => {
      return (document.definitions as OperationDefinitionNode[]).flatMap((op) => {
        return (op.variableDefinitions || []).map((v) => {
          return [`${index}.variables.${v.variable.name.value}`, v]
        })
      })
    }),
  )
}
