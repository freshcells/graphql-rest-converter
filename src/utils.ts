/* eslint-disable @typescript-eslint/ban-ts-comment */
import {
  BridgeOperations,
  CreateOpenAPISchemaConfig,
  CustomOperationProps,
  SchemaComponents,
} from './types.js'
import { createOpenAPISchemaFromOperations } from './openApi.js'
import OpenAPISchemaValidatorImport from 'openapi-schema-validator'
import { VariableDefinitionNode, VariableNode } from 'graphql'
import { getDirective, getDirectiveArgumentsUntyped } from './graphqlUtils.js'
import { OpenAPIDirectives } from './graphql.js'

// @ts-ignore
const OpenAPISchemaValidator = OpenAPISchemaValidatorImport.default || OpenAPISchemaValidatorImport

export const getParameterName = (node: VariableNode, nodeDef: VariableDefinitionNode) => {
  const paramDirectiveValue = getDirective(nodeDef, OpenAPIDirectives.Param)
  if (!paramDirectiveValue) {
    return node.name.value
  }
  const directiveParams = getDirectiveArgumentsUntyped(paramDirectiveValue)

  return (directiveParams['name'] as string) || node.name.value
}

export const resolveSchemaComponents = (schema: any, schemaComponents: SchemaComponents) => {
  if (typeof schema !== 'object' || schema === null) {
    return
  }
  if ('$ref' in schema) {
    const refTarget = schema.$ref as string
    delete schema.$ref
    for (const [k, v] of Object.entries(
      schemaComponents[refTarget.replace('#/components/schemas/', '')],
    )) {
      schema[k] = structuredClone(v)
    }
  }
  for (const value of Object.values(schema)) {
    resolveSchemaComponents(value, schemaComponents)
  }
}

export const createOpenAPISchemaWithValidate = <
  T extends CustomOperationProps = CustomOperationProps,
>(
  operations: BridgeOperations<T>,
  config?: CreateOpenAPISchemaConfig<T>,
) => {
  const openAPISchema = createOpenAPISchemaFromOperations<T>(config?.baseSchema || {}, operations)
  if (config?.validate) {
    const schemaValidator = new OpenAPISchemaValidator({ version: 3 })
    const schemaValidationErrors = schemaValidator.validate(openAPISchema)
    if (schemaValidationErrors?.errors?.length) {
      throw new Error(JSON.stringify(schemaValidationErrors))
    }
  }
  return openAPISchema
}
