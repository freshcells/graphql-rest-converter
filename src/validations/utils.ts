import {
  GraphQLInputType,
  GraphQLNonNull,
  GraphQLSchema,
  OperationDefinitionNode,
  typeFromAST,
  VariableDefinitionNode,
  VariableNode,
  GraphQLError,
  ValidationContext,
} from 'graphql'
import type { Maybe } from 'graphql/jsutils/Maybe.js'
import { inspect } from 'graphql/jsutils/inspect.js'
import { getParameterName } from '../utils.js'
import { getDirectiveArgumentsWithSchema, hasDirective } from '../graphqlUtils.js'
import { OpenAPIDirectives } from '../graphql.js'

export const validateVariable = (
  varDefMap: Record<string, VariableDefinitionNode>,
  context: ValidationContext,
  pathVariables: Set<string>,
  node: VariableNode,
  type: Maybe<GraphQLInputType>,
  operation: OperationDefinitionNode,
  pathDefinition: string,
) => {
  const varName = node.name.value
  const varDef = varDefMap[varName]
  if (!varDef) {
    return
  }
  const varType = typeFromAST(context.getSchema(), varDef.type)
  const varTypeStr = inspect(varType)

  const parameterName = getParameterName(node, varDef)
  const paramDirectiveArguments = getDirectiveArgumentsWithSchema(
    context.getSchema(),
    varDef,
    OpenAPIDirectives.Param,
  )
  const inValue = (paramDirectiveArguments?.in as string | undefined)?.toLowerCase()
  if (pathVariables.has(parameterName)) {
    if (hasDirective(varDef, OpenAPIDirectives.Body)) {
      context.reportError(
        new GraphQLError(
          `Variable "$${varName}" of type "${varTypeStr}" cannot be used with "@OABody", as it is used within "${pathDefinition}".`,
          {
            nodes: [varDef, node, operation],
          },
        ),
      )
      return
    }

    if (!(varType instanceof GraphQLNonNull)) {
      context.reportError(
        new GraphQLError(
          `Variable "$${varName}" of type "${varTypeStr}" must be defined as "${varTypeStr}!", as it is used within "${pathDefinition}".`,
          {
            nodes: [varDef, node, operation],
          },
        ),
      )
      return
    }
  }

  if (inValue && pathVariables.has(parameterName) && inValue !== 'path') {
    context.reportError(
      new GraphQLError(
        `Location "${inValue}" is invalid for "$${varName}" of type "${varTypeStr}", because "${parameterName}" is part of the path "${pathDefinition}".`,
        {
          nodes: [varDef, node, operation],
        },
      ),
    )
  }
  if (!pathVariables.has(parameterName) && inValue === 'path') {
    context.reportError(
      new GraphQLError(
        `Location "${inValue}" is invalid for "$${varName}" of type "${varTypeStr}", because "${parameterName}" was expected in "${pathDefinition}".`,
        {
          nodes: [varDef, node, operation],
        },
      ),
    )
  }
}

export const getBodyDirectiveInfo = (
  varDefMap: Record<string, VariableDefinitionNode>,
  node: VariableNode,
  schema: GraphQLSchema,
) => {
  const varName = node.name.value
  const varDef = varDefMap[varName]
  if (!varDef) {
    return null
  }
  const directiveArguments = getDirectiveArgumentsWithSchema(schema, varDef, OpenAPIDirectives.Body)
  return directiveArguments
    ? {
        varDef,
        node,
        path: directiveArguments.path || varName,
        contentType: directiveArguments.contentType,
      }
    : null
}
