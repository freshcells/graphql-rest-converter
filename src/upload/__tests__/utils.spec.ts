import { parse } from 'graphql'
import { getAllVariablesFromDocuments } from '../utils.js'

test('getAllVariablesFromDocuments', () => {
  const document1 = parse(`
    mutation upload($var1: Upload, $var2: Upload!, $var3: Upload) {
      upload(var1: $var1, var2: $var2, var3: $var3)
    }
  `)

  const result = getAllVariablesFromDocuments([document1, document1])

  expect(result).toMatchSnapshot()
})
