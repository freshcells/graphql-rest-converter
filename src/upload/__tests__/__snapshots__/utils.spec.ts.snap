// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAllVariablesFromDocuments 1`] = `
{
  "0.variables.var1": {
    "defaultValue": undefined,
    "directives": [],
    "kind": "VariableDefinition",
    "loc": {
      "end": 34,
      "start": 21,
    },
    "type": {
      "kind": "NamedType",
      "loc": {
        "end": 34,
        "start": 28,
      },
      "name": {
        "kind": "Name",
        "loc": {
          "end": 34,
          "start": 28,
        },
        "value": "Upload",
      },
    },
    "variable": {
      "kind": "Variable",
      "loc": {
        "end": 26,
        "start": 21,
      },
      "name": {
        "kind": "Name",
        "loc": {
          "end": 26,
          "start": 22,
        },
        "value": "var1",
      },
    },
  },
  "0.variables.var2": {
    "defaultValue": undefined,
    "directives": [],
    "kind": "VariableDefinition",
    "loc": {
      "end": 50,
      "start": 36,
    },
    "type": {
      "kind": "NonNullType",
      "loc": {
        "end": 50,
        "start": 43,
      },
      "type": {
        "kind": "NamedType",
        "loc": {
          "end": 49,
          "start": 43,
        },
        "name": {
          "kind": "Name",
          "loc": {
            "end": 49,
            "start": 43,
          },
          "value": "Upload",
        },
      },
    },
    "variable": {
      "kind": "Variable",
      "loc": {
        "end": 41,
        "start": 36,
      },
      "name": {
        "kind": "Name",
        "loc": {
          "end": 41,
          "start": 37,
        },
        "value": "var2",
      },
    },
  },
  "0.variables.var3": {
    "defaultValue": undefined,
    "directives": [],
    "kind": "VariableDefinition",
    "loc": {
      "end": 65,
      "start": 52,
    },
    "type": {
      "kind": "NamedType",
      "loc": {
        "end": 65,
        "start": 59,
      },
      "name": {
        "kind": "Name",
        "loc": {
          "end": 65,
          "start": 59,
        },
        "value": "Upload",
      },
    },
    "variable": {
      "kind": "Variable",
      "loc": {
        "end": 57,
        "start": 52,
      },
      "name": {
        "kind": "Name",
        "loc": {
          "end": 57,
          "start": 53,
        },
        "value": "var3",
      },
    },
  },
  "1.variables.var1": {
    "defaultValue": undefined,
    "directives": [],
    "kind": "VariableDefinition",
    "loc": {
      "end": 34,
      "start": 21,
    },
    "type": {
      "kind": "NamedType",
      "loc": {
        "end": 34,
        "start": 28,
      },
      "name": {
        "kind": "Name",
        "loc": {
          "end": 34,
          "start": 28,
        },
        "value": "Upload",
      },
    },
    "variable": {
      "kind": "Variable",
      "loc": {
        "end": 26,
        "start": 21,
      },
      "name": {
        "kind": "Name",
        "loc": {
          "end": 26,
          "start": 22,
        },
        "value": "var1",
      },
    },
  },
  "1.variables.var2": {
    "defaultValue": undefined,
    "directives": [],
    "kind": "VariableDefinition",
    "loc": {
      "end": 50,
      "start": 36,
    },
    "type": {
      "kind": "NonNullType",
      "loc": {
        "end": 50,
        "start": 43,
      },
      "type": {
        "kind": "NamedType",
        "loc": {
          "end": 49,
          "start": 43,
        },
        "name": {
          "kind": "Name",
          "loc": {
            "end": 49,
            "start": 43,
          },
          "value": "Upload",
        },
      },
    },
    "variable": {
      "kind": "Variable",
      "loc": {
        "end": 41,
        "start": 36,
      },
      "name": {
        "kind": "Name",
        "loc": {
          "end": 41,
          "start": 37,
        },
        "value": "var2",
      },
    },
  },
  "1.variables.var3": {
    "defaultValue": undefined,
    "directives": [],
    "kind": "VariableDefinition",
    "loc": {
      "end": 65,
      "start": 52,
    },
    "type": {
      "kind": "NamedType",
      "loc": {
        "end": 65,
        "start": 59,
      },
      "name": {
        "kind": "Name",
        "loc": {
          "end": 65,
          "start": 59,
        },
        "value": "Upload",
      },
    },
    "variable": {
      "kind": "Variable",
      "loc": {
        "end": 57,
        "start": 52,
      },
      "name": {
        "kind": "Name",
        "loc": {
          "end": 57,
          "start": 53,
        },
        "value": "var3",
      },
    },
  },
}
`;
